library/src/main/scala/scala/scalajs/js/Promise.scala

```diff
@@ -37,26 +37,35 @@ import scala.concurrent.Future
  *  directly use the methods of `Future` on `Promise`s.
  */
 @js.native
-class Promise[+A](
+class Promise[+A] private[js] (
     executor: js.Function2[js.Function1[A | Thenable[A], _], js.Function1[scala.Any, _], _])
     extends js.Object with js.Thenable[A] {
 
-  def `then`[B](
-      onFulfilled: js.Function1[A, B | Thenable[B]],
-      onRejected: js.UndefOr[js.Function1[scala.Any, B | Thenable[B]]] = js.undefined): Thenable[B] = js.native
+  def `then`[B, C](
+      onFulfilled: js.Function1[A, B],
+      onRejected: js.UndefOr[js.Function1[scala.Any, B]] = js.undefined
+  )(implicit
+      ev: Thenable.Returning[B, C]
```

Comment:
You can't do this. An actual value of `Returning` will be given as a third parameter when calling `then`. For some `Thenable`s, this could have an unwanted effect (if they use the third param for options or something).